msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1263,6 +1268,131 @@ public boolean isSharedDevice() {
+                account,
+                popParameters,
+                new CommandCallback<GenerateShrResult, BaseException>() {
+                    @Override
+                    public void onCancel() {
+                        // Not cancellable
+                    }
+
+                    @Override
+                    public void onError(@NonNull final BaseException error) {
+                        future.setResult(
+                                new AsyncResult<GenerateShrResult>(
+                                        null,
+                                        baseExceptionToMsalException(error)
+                                )
+                        );
+                    }
+
+                    @Override
+                    public void onTaskCompleted(@NonNull final GenerateShrResult generateShrResult) {
+                        future.setResult(new AsyncResult<>(generateShrResult, null));
+                    }
+                },
+                "1100" // Public API ID
+        );
+
+        // Execute this command silently...
+        CommandDispatcher.submitSilent(generateShrCommand);
+
+        try {
+            final AsyncResult<GenerateShrResult> asyncResult = future.get();
+
+            if (asyncResult.getSuccess()) {
+                return asyncResult.getResult().getShr();
+            } else {
+                throw asyncResult.getException();
+            }
+        } catch (ExecutionException | InterruptedException e) {
+            throw new MsalClientException(
+                    UNKNOWN_ERROR,
+                    "Unexpected error while generating SHR.",
+                    e
+            );
+        }
+    }
+
+    @Override
+    public void generateSignedHttpRequest(@NonNull final IAccount account,
+                                          @NonNull final PoPAuthenticationScheme popParameters,
+                                          @NonNull final SignedHttpRequestRequestCallback callback) {
+        try {
+            final GenerateShrCommand generateShrCommand = createGenerateShrCommand(
+                    account,
+                    popParameters,
+                    new CommandCallback<GenerateShrResult, BaseException>() {
+                        @Override
+                        public void onCancel() {
+                            // Not cancellable
+                        }
+
+                        @Override
+                        public void onError(@NonNull final BaseException error) {
+                            callback.onError(baseExceptionToMsalException(error));
+                        }
+
+                        @Override
+                        public void onTaskCompleted(@NonNull final GenerateShrResult generateShrResult) {
+                            callback.onTaskCompleted(generateShrResult.getShr());
+                        }
+                    },
+                    "1101" // Public API ID
```

Comment:
move this constant to `PublicApiId.java`

Reply:
cc33739f1c87ea2ae37c86ab5a15ef440c3a5b6e